src/main.rs

```diff
@@ -160,24 +179,62 @@ async fn write_state(infile: &str, outfile: &str, cachefile: &str) {
     geo_cache.save().await.expect("could not save geoip cache");
 
     let joutput = serde_json::to_string(&state).unwrap();
-    fs::write(outfile, joutput).unwrap();
+    fs::write(config.state_file_path.as_ref().unwrap(), joutput).unwrap();
 }
 
 #[tokio::main]
 async fn main() {
-    let args: Vec<String> = env::args().collect();
+    let arg_conf = ArgConfiguration::parse();
+    let mut configuration: CrunchyConfiguration;
 
-    // TODO(asmie): could be refactored to use eg. clap crate.
-    if args.len() != 4 {
-        println!("\n\nUsage is: cargo run <in-sample.json> <out-state.json> <geoip-cache.json> ");
-        println!("E.g.:     \x1b[93mcargo run --release testdata/sample.json testdata/state.json\x1b[0m\n");
-        return;
+    if let Some(config_file_path) = arg_conf.config_file {
+        configuration = CrunchyConfiguration::new(&config_file_path)
+            .expect("configuration file could not be loaded or parsed");
+    } else {
+        configuration = CrunchyConfiguration::default();
     }
-    if fs::metadata(&args[1]).is_err() {
-        println!("{}: No such file or directory", &args[1]);
+
+    // Override configuration with command line arguments if provided
+    if let Some(input_sample) = arg_conf.input_sample {
+        configuration.input_file_path = Some(PathBuf::from(input_sample));
+    }
+    if let Some(out_state) = arg_conf.out_state {
+        configuration.state_file_path = Some(PathBuf::from(out_state));
+    }
+    if let Some(geocache_file) = arg_conf.geocache_file {
+        configuration.geoip_config.geocache_file_path = PathBuf::from(geocache_file);
+    }
+
+    if fs::metadata(configuration.input_file_path.as_ref().unwrap()).is_err() {
```

Comment:
We don't really need metadata here, just checking if file/dir exists, right? Can this check be done with Path or PathBuf, simultaneously checking for is_file ?

Reply:
Changed.